xc/device.ml

```diff
@@ -1414,8 +1426,71 @@ module PCI = struct
       Forkhelpers.execute_command_get_output !Resources.rmmod ["i915"] in ()
 
   let procfs_nvidia = "/proc/driver/nvidia/gpus"
+  let nvidia_smi    = "/usr/bin/nvidia-smi"
+  let nvidia_manage = "/usr/lib/nvidia/sriov-manage"
+
+  (** [num_vfs devstr] returns the number of PCI VFs of [devstr] or 0 if
+    * [devstr] is not an SRIOV device
+    *)
+  let num_vfs devstr =
+    let path = Printf.sprintf "/sys/bus/pci/devices/%s/sriov_numvfs" devstr in
+    try
+      Unixext.string_of_file path |> String.trim |> int_of_string
+    with
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+        debug "File %s does not exist - assuming no SRIOV devices in use" path;
+        0
+      | exn ->
+        let msg = Printf.sprintf "Can't read %s to reset Nvidia GPU %s: %s"
+          path devstr (Printexc.to_string exn) in
+        error "%s" msg;
+        raise (Xenopsd_error (Internal_error msg))
+
+  (** [vfs_of device] returns the PCI addresses of the virtual functions
+   * of PCI [device]. We find each virtual function by looking at the
+   * virtfnX symlink in [device].
+   *)
+  let vfs_of devstr =
+    let device = Filename.concat sysfs_devices devstr in
+    let some x = Some x in
+    let virtfn n =
+      let path = Printf.sprintf "virtfn%d" n in
+      try
+        Filename.concat device path |> Unix.readlink |> Filename.basename |> some
+      with exn ->
+        debug "Can't read %s of PCI %s: %s" path devstr (Printexc.to_string exn);
+        None in
+    let rec seq = function
```

Comment:
could avoid explicit recursion with `List.init (n-1) (fun x -> (n-1) - x)`

Reply:
```

Reply:
the following works:

```
utop # let seq' n = List.init n (fun x -> max (n -1 -x) 0);;
utop # seq' 3;;
- : int list = [2; 1; 0]
utop # seq' 0;;
- : int list = []
```

Reply:
@psafont pointed out we don't need the max, since the minimum of `n-1-x` is 0 (when `x = n -1`)